### PR TITLE
Xray Ignore rules support

### PR DESCRIPTION
**Pull request by dortam888**
- [ ] All [tests](https://github.com/jfrog/jfrog-client-go#tests) passed. If this feature is not already covered by the tests, I added new tests.
- [x] All [static analysis checks](https://github.com/jfrog/jfrog-client-go/actions/workflows/analysis.yml) passed.
- [x] This pull request is on the dev branch.
- [x] I used gofmt for formatting the code before submitting the pull request.
-----

Adding Support For Xray Ignore rules for testing

### CODE REVIEW

File: xray/services/ignorerule.go [hunk 1]
```diff
@@ -0,0 +1,138 @@
...
+		return &utils.IgnoreRuleParams{}, err
+	}
+
+	err = json.Unmarshal(body, ignoreRule)
+
+	if err != nil {
+		return &utils.IgnoreRuleParams{}, errors.New("failed unmarshalling ignore rule " + ignoreRuleId)
+	}
+
+	log.Debug("Xray response:", resp.Status)
```
Comment (attiasas) [inline xray/services/ignorerule.go:130]:
```suggestion
	log.Debug("Xray response status:", resp.Status)
```

File: xray/services/utils/ignorerulebody.go [hunk 2]
```diff
@@ -0,0 +1,65 @@
...
+}
+
+type IgnoreRuleBody struct {
+	Id            string        `json:"id,omitempty"`
+	Author        string        `json:"author,omitempty"`
+	Created       time.Time     `json:"created,omitempty"`
+	IsExpired     bool          `json:"is_expired,omitempty"`
+	Notes         string        `json:"notes"`
+	ExpiresAt     time.Time     `json:"expires_at,omitempty"`
+	IgnoreFilters IgnoreFilters `json:"ignore_filters"`
```
Comment (attiasas) [inline xray/services/utils/ignorerulebody.go:18]:
```suggestion
	IgnoreRuleParams
```
Same as `IgnoreRuleParams`

File: tests/xrayignorerule_test.go [hunk 3]
```diff
@@ -0,0 +1,54 @@
...
+	}
+	components := []utils.IgnoreFilterNameVersion{component}
+
+	cve := []string{"CVE-2022-31197"}
+	ignoreRuleFilter := utils.IgnoreFilters{
+		CVEs:       cve,
+		Components: components,
+	}
+
+	createIgnoreRule(t, &ignoreRuleId, ignoreRuleFilter)
```
Comment (attiasas) [inline tests/xrayignorerule_test.go:41]: You are returning `*utils.IgnoreRuleParams` from `createIgnoreRule` but it is not used

**Review by attiasas — CHANGES_REQUESTED** [review]
Nice work! check out my comments.
In addition:
* Add changes to the `README.md` content

File: xray/services/utils/ignorerulebody.go [hunk 4]
```diff
@@ -0,0 +1,69 @@
...
+}
+
+func CreateIgnoreRuleBody(ignoreRuleParams IgnoreRuleParams) IgnoreRuleBody {
+	return IgnoreRuleBody{
+		IgnoreRuleParams: IgnoreRuleParams{
+			Notes:         ignoreRuleParams.Notes,
+			ExpiresAt:     ignoreRuleParams.ExpiresAt,
+			IgnoreFilters: ignoreRuleParams.IgnoreFilters,
+		},
+	}
```
Comment (attiasas) [inline xray/services/utils/ignorerulebody.go:68]:
```suggestion
	return IgnoreRuleBody{
		IgnoreRuleParams: IgnoreRuleParams,
	}
```
You can just assign the value

File: xray/services/utils/ignorerulebody.go [hunk 5]
```diff
@@ -0,0 +1,69 @@
...
+type IgnoreFilters struct {
+	Vulnerabilities  []string                      `json:"vulnerabilities,omitempty"`
+	Licenses         []string                      `json:"licenses,omitempty"`
+	CVEs             []string                      `json:"cves,omitempty"`
+	Policies         []string                      `json:"policies,omitempty"`
+	Watches          []string                      `json:"watches,omitempty"`
+	DockerLayers     []string                      `json:"docker-layers,omitempty"`
+	OperationalRisks []string                      `json:"operational_risk,omitempty"`
+	Exposures        []ExposuresFilterName         `json:"exposures,omitempty"`
+	ReleaseBundles   []IgnoreFilterNameVersion     `json:"release-bundles,omitempty"`
```
Comment (attiasas) [inline xray/services/utils/ignorerulebody.go:28]:
```suggestion
	ReleaseBundles   []IgnoreFilterNameVersion     `json:"release_bundles,omitempty"`
```
please use the same type of format cases and don't mix (you have snake-case and kebab-case, or is it what the API is expecting?

File: xray/services/utils/ignorerulebody.go [hunk 6]
```diff
@@ -0,0 +1,69 @@
...
+	IgnoreRuleParams
+}
+
+type IgnoreFilters struct {
+	Vulnerabilities  []string                      `json:"vulnerabilities,omitempty"`
+	Licenses         []string                      `json:"licenses,omitempty"`
+	CVEs             []string                      `json:"cves,omitempty"`
+	Policies         []string                      `json:"policies,omitempty"`
+	Watches          []string                      `json:"watches,omitempty"`
+	DockerLayers     []string                      `json:"docker-layers,omitempty"`
```
Comment (attiasas) [inline xray/services/utils/ignorerulebody.go:25]:
```suggestion
	DockerLayers     []string                      `json:"docker_layers,omitempty"`
```
please use the same type of format cases and don't mix (you have snake-case and kebab-case, or is it what the API is expecting?

File: xray/services/ignorerule.go [hunk 7]
```diff
@@ -0,0 +1,143 @@
...
+	}
+
+	log.Debug("Xray response status:", resp.Status)
+	log.Info("Done getting ignore rule.")
+
+	return &utils.IgnoreRuleParams{
+		Notes:         ignoreRule.Notes,
+		ExpiresAt:     ignoreRule.ExpiresAt,
+		IgnoreFilters: ignoreRule.IgnoreFilters,
+	}, nil
```
Comment (attiasas) [inline xray/services/ignorerule.go:142]:
```suggestion
	return &ignoreRule.IgnoreRuleParams, nil
```
You can access params directly

File: xray/services/ignorerule.go [hunk 8]
```diff
@@ -0,0 +1,143 @@
...
+	}
+
+	return match[1], nil
+}
+
+// Get retrieves the details about an Xray ignore rule by its id
+// It will error if no ignore rule can be found by that id.
+func (xirs *IgnoreRuleService) Get(ignoreRuleId string) (ignoreRuleResp *utils.IgnoreRuleParams, err error) {
+	httpClientsDetails := xirs.XrayDetails.CreateHttpClientDetails()
+	log.Info("Getting ignore rule '%s'...", ignoreRuleId)
```
Comment (attiasas) [inline xray/services/ignorerule.go:120]:
```suggestion
	log.Info(fmt.Spirntf("Getting ignore rule '%s'...", ignoreRuleId))
```
missing `fmt`

File: xray/services/utils/ignorerulebody.go (hunk 5 above)
Comment (dortam888) [reply xray/services/utils/ignorerulebody.go:28]: This is what the API expects. It failed the test I added on release bundle.
See [here](https://jfrog.com/help/r/xray-rest-apis/create-ignore-rule)  operational_risk is snake and release-bundles and docker-layers is kebab

**Comment by 3choBoomer** [comment]
duplicate of https://github.com/jfrog/jfrog-client-go/pull/960 ?

File: tests/xrayignorerule_test.go [hunk 9]
```diff
@@ -0,0 +1,130 @@
...
+	}
+	components := []utils.IgnoreFilterNameVersion{component}
+
+	cve := []string{"CVE-2022-31197"}
+	ignoreRuleFilter := utils.IgnoreFilters{
+		CVEs:       cve,
+		Components: components,
+	}
+
+	ignoreRuleId = createIgnoreRule(t, ignoreRuleFilter)
```
Comment (eyalbe4) [inline tests/xrayignorerule_test.go:44]: This comment applies to then other tests that create ignore rules as well - 
Shouldn't we check that the `ignoreRuleId` isn't empty?

File: tests/xrayignorerule_test.go [hunk 10]
```diff
@@ -0,0 +1,130 @@
...
+	ignoreRuleId, err := testsXrayIgnoreRuleService.Create(ignoreRuleParams)
+	assert.NoError(t, err)
+	return ignoreRuleId
+}
+
+func createDummyWatch(policyName string, watchName string) error {
+	err := createDummyPolicy(policyName)
+	if err != nil {
+		return err
+	}
```
Comment (eyalbe4) [inline tests/xrayignorerule_test.go:114]: Can be shortened to - 
```
	if err := createDummyPolicy(policyName); err != nil {
		return err
	}
```

File: tests/xrayignorerule_test.go [hunk 11]
```diff
@@ -0,0 +1,130 @@
...
+		},
+		Policies: []utils.AssignedPolicy{
+			{
+				Name: policyName,
+				Type: "security",
+			},
+		},
+	}
+	err = testsXrayWatchService.Create(params)
+	return err
```
Comment (eyalbe4) [inline tests/xrayignorerule_test.go:129]: Can be shortened to -
```
return testsXrayWatchService.Create(params)
```

File: xray/services/ignorerule.go [hunk 12]
```diff
@@ -0,0 +1,140 @@
...
+}
+
+func getIgnoreRuleIdFromBody(body []byte) (string, error) {
+	str := string(body)
+
+	re := regexp.MustCompile(`id:\s*([a-f0-9-]+)`)
+	match := re.FindStringSubmatch(str)
+
+	if len(match) <= 1 {
+		err := errors.New("couldn't find id for ignore rule")
```
Comment (eyalbe4) [inline xray/services/ignorerule.go:110]: I recommend adding a more specific error message tp the error. Perhaps include the string that failed to be matched to the error.
In addition, this error should be wrapped with CheckError. See other examples in the code for this.

File: xray/services/ignorerule.go [hunk 13]
```diff
@@ -0,0 +1,140 @@
...
+		return "", err
+	}
+
+	return match[1], nil
+}
+
+// Get retrieves the details about an Xray ignore rule by its id
+// It will error if no ignore rule can be found by that id.
```
Comment (eyalbe4) [inline xray/services/ignorerule.go:118]: can --> cannot

File: xray/services/ignorerule.go [hunk 14]
```diff
@@ -0,0 +1,140 @@
...
+
+	if err != nil {
+		return &utils.IgnoreRuleParams{}, err
+	}
+	if err = errorutils.CheckResponseStatusWithBody(resp, body, http.StatusOK); err != nil {
+		return &utils.IgnoreRuleParams{}, err
+	}
+
+	if err = json.Unmarshal(body, ignoreRule); err != nil {
+		return &utils.IgnoreRuleParams{}, errors.New("failed unmarshalling ignore rule " + ignoreRuleId)
```
Comment (eyalbe4) [inline xray/services/ignorerule.go:133]: Missing CheckError here.

**Review by eyalbe4 — CHANGES_REQUESTED** [review]

**Comment by dortam888** [comment]
@3choBoomer Apologies, I didn't notice #960 when creating this PR. 
My PR is a bit more updated to the current API with exposures feature addition and some small fixes compared to yours (For example docker-layers and release-bundles). 
We will soon add more features to ignore rules regarding SAST, our team will support it as well as soon as the changes are released.
Again, apologies for overriding your PR there was no intention to do it and we happily accept all contributions.

File: tests/xrayignorerule_test.go (hunk 9 above)
Comment (dortam888) [reply tests/xrayignorerule_test.go:42]: I had the check at defer function and changed it to be in the function body with assertion.
If the ignoreRuleId is empty I would expect an error to occur earlier at the API phase but I don't want to ignore it in the test as it is an error.

File: tests/xrayignorerule_test.go (hunk 10 above)
Comment (dortam888) [reply tests/xrayignorerule_test.go]: done

File: tests/xrayignorerule_test.go (hunk 11 above)
Comment (dortam888) [reply tests/xrayignorerule_test.go]: done

File: xray/services/ignorerule.go (hunk 14 above)
Comment (dortam888) [reply xray/services/ignorerule.go]: Added CheckError

File: xray/services/ignorerule.go [hunk 15]
```diff
@@ -0,0 +1,138 @@
...
+
+	if err != nil {
+		return &utils.IgnoreRuleParams{}, err
+	}
+	if err = errorutils.CheckResponseStatusWithBody(resp, body, http.StatusOK); err != nil {
+		return &utils.IgnoreRuleParams{}, err
+	}
+
+	if err = json.Unmarshal(body, ignoreRule); err != nil {
+		return &utils.IgnoreRuleParams{}, errorutils.CheckErrorf("failed unmarshalling ignore rule %s", ignoreRuleId)
```
Comment (eyalbe4) [inline xray/services/ignorerule.go:131]: I recommend adding the value of `string(body)` to the error message.

**Review by eyalbe4 — CHANGES_REQUESTED** [review]

**Pull request merged**